hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/ResourceCalculationDriver.java

```diff
@@ -0,0 +1,368 @@
+  protected CSQueue currentChild;
+  protected Map<String, Float> usedResourceByCurrentCalculator = new HashMap<>();
+
+  public ResourceCalculationDriver(
+      CSQueue parent, QueueCapacityUpdateContext updateContext,
+      Map<ResourceUnitCapacityType, AbstractQueueCapacityCalculator> calculators,
+      Collection<String> definedResources) {
+    this.parent = parent;
+    this.updateContext = updateContext;
+    this.calculators = calculators;
+    this.definedResources = definedResources;
+  }
+
+  /**
+   * Returns the parent that is driving the calculation.
+   *
+   * @return a common parent queue
+   */
+  public CSQueue getParent() {
+    return parent;
+  }
+
+  /**
+   * Returns the context that is used throughout the whole update phase.
+   *
+   * @return update context
+   */
+  public QueueCapacityUpdateContext getUpdateContext() {
+    return updateContext;
+  }
+
+  /**
+   * Returns the name of the resource that is currently processed.
+   *
+   * @return resource name
+   */
+  public String getCurrentResourceName() {
+    return currentResourceName;
+  }
+
+  /**
+   * Returns the child that is currently processed.
+   *
+   * @return child queue
+   */
+  public CSQueue getCurrentChild() {
+    return currentChild;
+  }
+
+  /**
+   * Sets the currently evaluated child to a specific queue.
+   *
+   * @param currentChild a child queue
+   */
+  public void setCurrentChild(CSQueue currentChild) {
+    if (currentChild.getParent() != parent) {
+      throw new IllegalArgumentException("Child queue " + currentChild.getQueuePath() + " is not " +
+          "a child of " + parent.getQueuePath());
+    }
+
+    this.currentChild = currentChild;
+  }
+
+  /**
+   * A shorthand to return the minimum capacity vector entry for the currently evaluated child and
+   * resource name.
+   *
+   * @param label node label
+   * @return capacity vector entry
+   */
+  public QueueCapacityVectorEntry getCurrentMinimumCapacityEntry(String label) {
+    return currentChild.getConfiguredCapacityVector(label).getResource(currentResourceName);
+  }
+
+  /**
+   * A shorthand to return the maximum capacity vector entry for the currently evaluated child and
+   * resource name.
+   *
+   * @param label node label
+   * @return capacity vector entry
+   */
+  public QueueCapacityVectorEntry getCurrentMaximumCapacityEntry(String label) {
+    return currentChild.getConfiguredMaxCapacityVector(label).getResource(currentResourceName);
+  }
+
+  /**
+   * Increments the aggregated weight.
+   *
+   * @param label        node label
+   * @param resourceName resource unit name
+   * @param value        weight value
+   */
+  public void incrementWeight(String label, String resourceName, float value) {
+    sumWeightsPerLabel.putIfAbsent(label, new HashMap<>());
+    sumWeightsPerLabel.get(label).put(resourceName,
+        sumWeightsPerLabel.get(label).getOrDefault(resourceName, 0f) + value);
```

Comment:
If this is called frequently we might consider using a mutable float here to reduce GC strain.

Reply:
@9uapaw Is this comment fixed?